src/test/unit/model/indexing/lvalue_varmat_test.cpp

```diff
@@ -0,0 +1,760 @@
+    EXPECT_MATRIX_EQ(x_val_check, x_rev_val_check);
+    sum(x).grad();
+    auto check_i = [i](int kk) { return kk <= i; };
+    auto check_j = [i](int jj) { return jj <= i; };
+    check_matrix_adjs(check_i, check_j, x, "lhs", 0);
+    check_matrix_adjs(check_i, check_j, x_rev, "rhs", 1);
+    stan::math::recover_memory();
+  }
+}
+
+TEST_F(VarAssign, minmax_uni_matrix) {
+  using stan::math::sum;
+  using stan::math::var_value;
+  using stan::model::test::check_matrix_adjs;
+  using stan::model::test::generate_linear_var_matrix;
+  using stan::model::test::generate_linear_var_vector;
+
+  auto x = generate_linear_var_matrix(3, 4);
+  Eigen::MatrixXd x_val = x.val();
+  auto y = generate_linear_var_vector(2, 10);
+
+  assign(x, index_list(index_min_max(2, 3), index_uni(4)), y);
+  EXPECT_MATRIX_EQ(y.val(), x.val().col(3).segment(1, 2));
+  sum(x).grad();
+  EXPECT_MATRIX_EQ(x.val(), x_val);
+  auto check_i = [](int i) { return (i == 1 || i == 2); };
+  auto check_j = [](int j) { return j == 3; };
+  check_matrix_adjs(check_i, check_j, x, "lhs", 0);
+  EXPECT_FLOAT_EQ(y.adj()(0), 1);
+  EXPECT_FLOAT_EQ(y.adj()(1), 1);
+
+  test_throw_out_of_range(x, index_list(index_min_max(2, 3), index_uni(0)), y);
+  test_throw_out_of_range(x, index_list(index_min_max(2, 3), index_uni(5)), y);
+  test_throw_out_of_range(x, index_list(index_min_max(0, 1), index_uni(4)), y);
+  test_throw_invalid_arg(x, index_list(index_min_max(1, 3), index_uni(4)), y);
+}
+
+TEST_F(VarAssign, minmax_min_matrix) {
+  using stan::math::sum;
+  using stan::math::var_value;
+  using stan::model::test::check_matrix_adjs;
+  using stan::model::test::generate_linear_var_matrix;
+
+  auto x = generate_linear_var_matrix(3, 4);
+  Eigen::MatrixXd x_val = x.val();
+  auto y = generate_linear_var_matrix(2, 3, 10);
+  assign(x, index_list(index_min_max(2, 3), index_min(2)), y);
+  EXPECT_MATRIX_EQ(y.val(), x.val().block(1, 1, 2, 3));
+  sum(x).grad();
+  EXPECT_MATRIX_EQ(x.val(), x_val);
+  auto check_i = [](int i) { return (i == 1 || i == 2); };
+  auto check_j = [](int j) { return j > 0; };
+  check_matrix_adjs(check_i, check_j, x, "lhs", 0);
+  EXPECT_MATRIX_EQ(y.adj(), MatrixXd::Ones(2, 3));
+}
+
+// omni assigns
+TEST_F(VarAssign, omni_uni_matrix) {
+  using stan::math::var_value;
+  using stan::model::test::check_matrix_adjs;
+  using stan::model::test::generate_linear_var_matrix;
+  using stan::model::test::generate_linear_var_vector;
+
+  auto x = generate_linear_var_matrix(5, 5);
+  Eigen::MatrixXd x_val = x.val();
+  auto y = generate_linear_var_vector<Eigen::VectorXd>(5, 10);
+  assign(x, index_list(index_omni(), index_uni(1)), y);
+  EXPECT_MATRIX_EQ(y.val(), x.val().col(0));
+  sum(x).grad();
+  EXPECT_MATRIX_EQ(x.val(), x_val);
+  auto check_i = [](int i) { return true; };
+  auto check_j = [](int j) { return j == 0; };
+  check_matrix_adjs(check_i, check_j, x, "lhs", 0);
+  EXPECT_MATRIX_EQ(y.adj(), Eigen::VectorXd::Ones(5));
+}
```

Comment:
I like the organization of tests. However I think some index combinations are missing. For example matrix with min and omni index. All allowed index combinations should be tested.

Reply:
Ooof that's a lot. What if here I did all single index and each one being on the right side of the double index? Doing all the combinations manually would be a lot of testing

Reply:
I added the above special case tests. It's not 100% converage but I think it captures a lot

Reply:
But we do support all index combinations. That means not everything we support is tested. I don't like that. 

Although writing 63 tests by hand really is a lot. Maybe it is possible to do something smart using templates or macros? For example instead of writing expected values by hand, you could write code for converting every index type into equivalent omni index. Than only omni omni combination needs to be tested by hand and all others can be compared to omni omni.

Reply:
But for the double index tests we are doing the specialization for the index on the right hand side and the specialization for the index as if it was just grabbing the rows. So while the way I have it now doesn't do all the cases doing the single index specialization and double index specialization does check each path. I'm not really understanding what edge cases we would capture by writing out each one individually

Reply:
Well for example an "edge case" of someone optimizing by adding a separate overload for a specific combination of indices. Also that way you test that all combinations actually can be instantiated.